crates/rustc_codegen_spirv/src/abi.rs

```diff
@@ -818,6 +818,29 @@ fn trans_image<'tcx>(
             }
             Some(SpirvType::Sampler.def(span, cx))
         }
+        SpirvAttribute::SampledImage => {
+            // see SpirvType::sizeof
+            if ty.size != Size::from_bytes(4) {
+                cx.tcx
+                    .sess
+                    .err("#[spirv(sampled_image)] type must have size 4");
+                return None;
+            }
+
+            // We use a generic to indicate the underlying image type of the sampled image.
+            // The spirv type of it will be generated by querying the type of the first generic.
+            if let TyKind::Adt(_, substs) = *ty.ty.kind() {
+                // TODO: enforce that the generic param is an image type?
+                let image_ty = substs.type_at(0);
+                let image_type = trans_type_impl(cx, span, cx.layout_of(image_ty), false);
+                Some(SpirvType::SampledImage { image_type }.def(span, cx))
+            } else {
+                cx.tcx
+                    .sess
+                    .err("#[spirv(sampled_image)] generic type must be an image");
```

Comment:
Nit: I think this might actually be a `bug!()`? I don't think it's possible to get a non-ADT here (see callsite of `trans_image`, where it's already matching on `TyKind::Adt` - I guess you could grab out the `substs` there and pass them into `trans_image`, if you'd like)